upup/pkg/fi/cloudup/populate_cluster_spec.go

```diff
@@ -217,6 +221,47 @@ func (c *populateClusterSpec) run(ctx context.Context, clientset simple.Clientse
 		klog.V(2).Infof("Normalizing kubernetes version: %q -> %q", cluster.Spec.KubernetesVersion, versionWithoutV)
 		cluster.Spec.KubernetesVersion = versionWithoutV
 	}
+
+	if cluster.Spec.CloudProvider.Openstack == nil {
+		if cluster.Spec.API.DNS == nil && cluster.Spec.API.LoadBalancer == nil {
+			subnetTypesByName := map[string]kopsapi.SubnetType{}
+			for _, subnet := range cluster.Spec.Networking.Subnets {
+				subnetTypesByName[subnet.Name] = subnet.Type
+			}
+
+			haveAPIServerNodes := false
+			for _, ig := range c.InputInstanceGroups {
+				if ig.Spec.Role == kopsapi.InstanceGroupRoleAPIServer {
+					haveAPIServerNodes = true
+				}
+			}
+			for _, ig := range c.InputInstanceGroups {
+				if ig.Spec.Role == kopsapi.InstanceGroupRoleAPIServer || (!haveAPIServerNodes && ig.Spec.Role == kopsapi.InstanceGroupRoleControlPlane) {
+					for _, subnet := range ig.Spec.Subnets {
+						switch subnetTypesByName[subnet] {
+						case kopsapi.SubnetTypePrivate:
+							if cluster.Spec.API.LoadBalancer == nil {
+								cluster.Spec.API.LoadBalancer = &kopsapi.LoadBalancerAccessSpec{}
+							}
+						case kopsapi.SubnetTypePublic:
+							if cluster.Spec.API.DNS == nil {
```

Comment:
```suggestion
							if cluster.Spec.API.LoadBalancer == nil && cluster.Spec.API.DNS == nil {
```

Reply:
Actually, either `if` condition will always be true due to the enclosing `if` statement.